x/application/keeper/msg_server_undelegate_from_gateway_test.go

```diff
@@ -272,3 +281,225 @@ func TestMsgServer_UndelegateFromGateway_SuccessfullyUndelegateFromUnstakedGatew
+}
+
+func TestMsgServer_UndelegateFromGateway_RedelegationAfterUndelegationAtTheSameSessionNumber(t *testing.T) {
+	k, ctx := keepertest.ApplicationKeeper(t)
+	srv := keeper.NewMsgServerImpl(k)
+
+	undelegationHeight := int64(1)
+	sdkCtx, app, gatewayAddrs := delegateThenUndelegate(ctx, t, srv, k, undelegationHeight)
+
+	gatewayAddrToRedelegate := gatewayAddrs[0]
+
+	// Increment the block height without moving to the next session.
+	sdkCtx = sdkCtx.WithBlockHeight(undelegationHeight + 1)
+
+	// Delegate back the application to the gateway that was undelegated from.
+	delegateMsg := &types.MsgDelegateToGateway{
+		AppAddress:     app.Address,
+		GatewayAddress: gatewayAddrToRedelegate,
+	}
+	_, err := srv.DelegateToGateway(ctx, delegateMsg)
+	require.NoError(t, err)
+
+	// Get the updated application state.
+	app, isAppFound := k.GetApplication(sdkCtx, app.Address)
+	require.True(t, isAppFound)
+	require.NotNil(t, app)
+
+	// Verify that the gateway is still in the application's delegatee gateway addresses.
+	require.Contains(t, app.DelegateeGatewayAddresses, gatewayAddrToRedelegate)
+
+	// Verify that the gateway is also present in the pending undelegation list with the
+	// right sessionEndHeight as the map key.
+	sessionEndHeight := uint64(sessionkeeper.GetSessionEndBlockHeight(undelegationHeight))
+	require.Contains(t,
+		app.PendingUndelegations[sessionEndHeight].GatewayAddresses,
+		gatewayAddrToRedelegate,
+	)
+
+	// Verify that the reconstructed delegatee gateway list include the redelegated gateway.
+	gatewayAddresses := rings.GetRingAddressesAtBlock(&app, sdkCtx.BlockHeight())
+	require.Contains(t, gatewayAddresses, gatewayAddrToRedelegate)
+
+	// Increment the block height past the undelegation retention period then run
+	// the pruning undelegations logic.
+	pruningBlockHeight := getPruningBlockHeight(undelegationHeight)
+	sdkCtx = sdkCtx.WithBlockHeight(pruningBlockHeight)
+	k.EndBlockerPruneAppToGatewayPendingUndelegation(sdkCtx)
+
+	// Get the updated application state after pruning logic is run.
+	app, isAppFound = k.GetApplication(sdkCtx, app.Address)
+	require.True(t, isAppFound)
+	require.NotNil(t, app)
+
+	// Verify that the application is still delegated to the gateway
+	require.Contains(t, app.DelegateeGatewayAddresses, gatewayAddrToRedelegate)
+
+	// Verify that the the pending undelegation map no longer contains the
+	// sessionEndHeight key.
+	require.Empty(t, app.PendingUndelegations[sessionEndHeight])
+
+	// Verify that the reconstructed delegatee gateway list includes the reundelegated gateway
+	gatewayAddressesAfterPruning := rings.GetRingAddressesAtBlock(&app, sdkCtx.BlockHeight())
+	require.Contains(t, gatewayAddressesAfterPruning, gatewayAddrToRedelegate)
+}
+
+// delegateThenUndelegate factors delegating an application to two gateways,
+// then undelegates the application from one of the gateways to be used in
+// multiple tests.
+func delegateThenUndelegate(
```

Comment:
Please update the comments and the name of the function to reflect all the "side effects" of this.

It does A LOT like creating an app, staking it, etc...